src/lustre/lustreRemoveMultAssign.ml

```diff
@@ -0,0 +1,299 @@
+  Example: 
+   if cond
+   then 
+      y1, y2 = node(expr1);
+   else
+      y1 = expr2;
+      y2 = expr3;
+   fi
+  -->
+   t1, t2 = node(expr1);
+   if cond
+   then 
+      y1 = t1;
+      y2 = t2;
+   else
+      y1 = expr2;
+      y2 = expr3;
+   fi
+
+  For each temp variable, we also generate a new declaration.
+*)
+
+module R = Res
+module A = LustreAst
+module AH = LustreAstHelpers
+module Ctx = TypeCheckerContext
+module Chk = LustreTypeChecker
+
+let (let*) = R.(>>=)
+
+type error_kind =
+  | MisplacedNodeItemError of A.node_item
+
+let error_message error = match error with
+  | MisplacedNodeItemError ni -> (match ni with
+    | Body (Assert _) -> "Asserts are not allowed inside if blocks or frame blocks."
+    | FrameBlock _ -> "Frame blocks are not allowed inside if blocks or frame blocks."
+    | AnnotMain _ -> "Main annotations are not allowed inside if blocks or frame blocks."
+    | AnnotProperty _ -> "Property annotations are not allowed inside if blocks or frame blocks."
+    (* Other node items are allowed *)
+    | _ -> assert false
+  )
+
+type error = [
+  | `LustreRemoveMultAssignError of Lib.position * error_kind
+  | `LustreAstInlineConstantsError of Lib.position * LustreAstInlineConstants.error_kind
+  | `LustreSyntaxChecksError of Lib.position * LustreSyntaxChecks.error_kind
+  | `LustreTypeCheckerError of Lib.position * LustreTypeChecker.error_kind
+]
+
+let mk_error pos kind = Error (`LustreRemoveMultAssignError (pos, kind))
+
+(** [i] is module state used to guarantee newly created identifiers are unique *)
+let i = ref (0)
+
+let split_and_flatten4 ls =
+  let xs = List.map (fun (x, _, _, _) -> x) ls |> List.flatten in
+  let ys = List.map (fun (_, y, _, _) -> y) ls |> List.flatten in
+  let zs = List.map (fun (_, _, z, _) -> z) ls |> List.flatten in
+  let ws = List.map (fun (_, _, _, w) -> w) ls |> List.flatten in
+  xs, ys, zs, ws
+
+(** Create a new fresh temporary variable name. *)
+let mk_fresh_temp_name _ =
+  i := !i + 1;
+  let prefix = HString.mk_hstring (string_of_int !i) in
+  HString.concat2 prefix (HString.mk_hstring "_temp") 
+
+(** When pulling out temp variables for recursive array definitions,
+   we also have to modify the RHS to match the temp variable.
+   For example, we want equations of the form
+   'temp[i] = if i = 0 then 0 else temp[i-1] + 1' rather than
+   'temp[i] = if i = 0 then 0 else y[i-1] + 1', where 'y' was
+   the original LHS variable name.
+*)
+let rec modify_arraydefs_in_expr arraydefs_original arraydefs_new = function
+    (* Replace all oracles with 'fill' *)
+    | A.Ident (pos, i1) -> 
+      let comb = List.combine arraydefs_original arraydefs_new in
+      let update = Lib.find_map (fun pair -> match pair with 
+        | (A.ArrayDef(_, i2, _), A.ArrayDef (_, i3, _)) when i1 = i2 -> 
+          Some (A.Ident(pos, i3))
+        | _ -> None
+      ) comb in (match update with 
+        | Some value -> value 
+        | None -> A.Ident(pos, i1)
+      )
```

Comment:
Notice that you are combining `arraydefs_original` and `arraydefs_new` for each identifier despite their combination does not depend on the identifier. This suggests you should compute their combination once before calling `modify_arraydefs_in_expr` and pass it to the function. Moreover, the call to `Lib.find_map` is basically using `comb` as an association list. So I suggest to replace this code with the following code:
```
| A.Ident (pos, i1) -> (
  let update = List.assoc_opt i1 array_assoc_list in
  match update with 
  | Some id -> A.Ident(pos, id)
  | None -> A.Ident(pos, i1)
)
```
where `array_assoc_list` is a new argument that replaces arguments `arraydefs_original` and `arraydefs_new`, and can be computed in `create_new_eqs` before the call to `modify_arraydefs_in_expr`  as follows:
```
let get_array_ids =
   List.filter_map (function
     | A.ArrayDef (_, id, _) -> Some id
     | _ -> None)
in
let array_assoc_list =
  let arrayids_original = get_array_ids ss in
  let arrayids_new = get_array_ids sis in
  List.combine arrayids_original arrayids_new
in
let expr = modify_arraydefs_in_expr array_assoc_list expr in
```